### ISSUE TITLE
[Project]: Add Real-Time Deepfake Audio Detection using Deep Learning (Wav2Vec2 + Binary Classifier)

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1197]
## Summary

I would like to propose and implement a new project for DL-Simplified: **Real-Time Deepfake Audio Detection using Deep Learning.** With the proliferation of AI voice cloning tools, the ability to detect synthetically generated speech has become a critical security problem. This project would train a binary classifier on top of the `Wav2Vec2` pre-trained speech model to distinguish real human speech from AI-generated (deepfake) audio. This fits the repository's scope as an advanced-level DL project and introduces contributors to audio deep learning, a significantly underrepresented domain in the current project collection.

## Why This Project

Looking at the current DL-Simplified project list, the audio/speech domain is extremely sparse (only `Cats vs Dogs Audio Classification` exists). The majority of projects are computer vision or NLP. Adding a cutting-edge, real-world-impact audio classification project meaningfully diversifies the repository's scope and introduces contributors to:
- Mel-spectrogram feature engineering
- Pre-trained Transformer-based speech models (`Wav2Vec2`)
- Binary classification fine-tuning
- Audio data augmentation techniques

## Proposed Implementation

**Dataset:** ASVspoof 2019 Dataset (publicly available, academic license) — contains 121,461 utterances split between genuine and spoofed speech across 17 different TTS/VC spoofing algorithms.

**Model Architecture:**
```python
import torch
import torch.nn as nn
from transformers import Wav2Vec2Model

class DeepfakeAudioDetector(nn.Module):
    def __init__(self, pretrained_model="facebook/wav2vec2-base"):
        super().__init__()
        self.wav2vec2 = Wav2Vec2Model.from_pretrained(pretrained_model)
        # Freeze lower 6 transformer layers; fine-tune top 6
        for i, layer in enumerate(self.wav2vec2.encoder.layers):
            if i < 6:
                for param in layer.parameters():
                    param.requires_grad = False
        self.classifier = nn.Sequential(
            nn.Linear(768, 256),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(256, 64),
            nn.ReLU(),
            nn.Linear(64, 1),
            nn.Sigmoid()
        )

    def forward(self, input_values, attention_mask=None):
        outputs = self.wav2vec2(input_values, attention_mask=attention_mask)
        # Mean pooling over sequence dimension
        hidden_states = outputs.last_hidden_state.mean(dim=1)
        return self.classifier(hidden_states).squeeze(-1)
```

**Training:**
```python
from torch.optim import AdamW
from torch.optim.lr_scheduler import CosineAnnealingLR

optimizer = AdamW(model.parameters(), lr=2e-5, weight_decay=0.01)
scheduler = CosineAnnealingLR(optimizer, T_max=10)
criterion = nn.BCELoss()
```

**Expected Results (based on published literature):**
- Equal Error Rate (EER): < 5% on ASVspoof 2019 eval set
- Accuracy: ~95–97% on held-out test split
- F1 Score: > 0.94

**Notebook structure (following the repo's template):**

Deepfake Audio Detection using DL/
├── Dataset/
│ └── README.md # ASVspoof 2019 download link + preprocessing steps
├── Images/
│ ├── mel_spectrogram_real.png
│ ├── mel_spectrogram_fake.png
│ ├── roc_curve.png
│ ├── confusion_matrix.png
│ └── training_curves.png
├── Model/
│ ├── Deepfake_Audio_Detection.ipynb
│ └── README.md
└── requirements.txt

**Libraries:** `transformers`, `datasets`, `torchaudio`, `librosa`, `scikit-learn`, `matplotlib`, `seaborn`

## Additional Notes

- Three models will be compared: (1) CNN on mel-spectrograms, (2) LSTM on MFCCs, (3) Wav2Vec2 fine-tuned — providing a learning progression from beginner to advanced.
- The README will include a section explaining the real-world significance: phone scam detection, identity verification, media authentication.
- A Gradio web app demo will be added under `Web App/` for interactive testing.

Could you please assign this issue to me?

**Labels:** `enhancement`, `new-project`, `audio-deep-learning`, `advanced`, `help wanted`, `GSSoC 2026`
